File: backend/youtube.py
from datetime import datetime, timedelta
from fastapi import HTTPException
from sqlalchemy.orm import Session
import googleapiclient.discovery
from googleapiclient.errors import HttpError

from auth import get_credentials


def get_youtube_client(user_id: str, db: Session):
    creds = get_credentials(user_id, db)
    return googleapiclient.discovery.build("youtube", "v3", credentials=creds)


def get_analytics_client(user_id: str, db: Session):
    creds = get_credentials(user_id, db)
    return googleapiclient.discovery.build("youtubeAnalytics", "v2", credentials=creds)


def get_recent_videos(user_id: str, db: Session, max_results: int = 50):
    youtube = get_youtube_client(user_id, db)

    # Get channel ID
    channels_response = youtube.channels().list(part="id", mine=True).execute()
    if not channels_response.get("items"):
        raise HTTPException(status_code=404, detail="No YouTube channel found")

    channel_id = channels_response["items"][0]["id"]

    # Get uploads playlist ID
    channel_details = (
        youtube.channels()
        .list(part="contentDetails", id=channel_id)
        .execute()
    )
    uploads_playlist_id = channel_details["items"][0]["contentDetails"][
        "relatedPlaylists"
    ]["uploads"]

    # Get playlist items
    playlist_items = []
    next_page_token = None

    while len(playlist_items) < max_results:
        request = youtube.playlistItems().list(
            part="contentDetails",
            playlistId=uploads_playlist_id,
            maxResults=min(50, max_results - len(playlist_items)),
            pageToken=next_page_token,
        )
        response = request.execute()
        playlist_items.extend(response.get("items", []))
        next_page_token = response.get("nextPageToken")
        if not next_page_token:
            break

    if not playlist_items:
        return []

    # Get video IDs
    video_ids = [
        item["contentDetails"]["videoId"] for item in playlist_items
    ]

    # Get video details in batches of 50
    videos = []
    for i in range(0, len(video_ids), 50):
        batch = video_ids[i : i + 50]
        videos_response = (
            youtube.videos()
            .list(
                part="snippet,statistics,contentDetails",
                id=",".join(batch),
            )
            .execute()
        )
        videos.extend(videos_response.get("items", []))

    # Format video data
    formatted_videos = []
    for video in videos:
        snippet = video.get("snippet", {})
        statistics = video.get("statistics", {})
        content_details = video.get("contentDetails", {})

        formatted_videos.append(
            {
                "id": video["id"],
                "title": snippet.get("title", ""),
                "description": snippet.get("description", ""),
                "tags": snippet.get("tags", []),
                "publishedAt": snippet.get("publishedAt", ""),
                "thumbnails": snippet.get("thumbnails", {}),
                "viewCount": int(statistics.get("viewCount", 0)),
                "likeCount": int(statistics.get("likeCount", 0)),
                "commentCount": int(statistics.get("commentCount", 0)),
                "duration": content_details.get("duration", ""),
            }
        )

    return formatted_videos


def get_video_analytics(user_id: str, db: Session, video_id: str, start_date: str = None):
    analytics = get_analytics_client(user_id, db)

    if not start_date:
        start_date = (datetime.now() - timedelta(days=90)).strftime("%Y-%m-%d")

    end_date = datetime.now().strftime("%Y-%m-%d")

    try:
        response = (
            analytics.reports()
            .query(
                ids="channel==MINE",
                startDate=start_date,
                endDate=end_date,
                metrics="estimatedMinutesWatched,averageViewDuration,averageViewPercentage,impressions,impressionClickThroughRate,subscribersGained",
                dimensions="video",
                filters=f"video=={video_id}",
            )
            .execute()
        )

        rows = response.get("rows", [])
        if not rows:
            return {}

        col_headers = [h["name"] for h in response.get("columnHeaders", [])]
        row = rows[0]

        return dict(zip(col_headers, row))
    except HttpError:
        return {}


def get_channel_analytics(user_id: str, db: Session):
    analytics = get_analytics_client(user_id, db)

    start_date = (datetime.now() - timedelta(days=90)).strftime("%Y-%m-%d")
    end_date = datetime.now().strftime("%Y-%m-%d")

    try:
        response = (
            analytics.reports()
            .query(
                ids="channel==MINE",
                startDate=start_date,
                endDate=end_date,
                metrics="views,estimatedMinutesWatched,subscribersGained,subscribersLost",
                dimensions="day",
            )
            .execute()
        )

        col_headers = [h["name"] for h in response.get("columnHeaders", [])]
        rows = response.get("rows", [])

        daily_data = []
        for row in rows:
            daily_data.append(dict(zip(col_headers, row)))

        # Aggregate totals
        totals = {
            "views": sum(int(r.get("views", 0)) for r in daily_data),
            "watchTime": sum(float(r.get("estimatedMinutesWatched", 0)) for r in daily_data),
            "subscribersGained": sum(int(r.get("subscribersGained", 0)) for r in daily_data),
            "subscribersLost": sum(int(r.get("subscribersLost", 0)) for r in daily_data),
            "dailyData": daily_data,
        }

        return totals
    except HttpError as e:
        raise HTTPException(status_code=500, detail=f"Analytics API error: {str(e)}")

File: apps/api/src/platforms/youtube.ts
import { google } from 'googleapis'
import type { NormalizedPost, PostMetrics, ChannelAnalytics } from '@analyzer/types'
import { BasePlatform } from './base.js'
import { getPlatformTokens, updatePlatformTokens } from '../db/index.js'

function makeOAuth2Client() {
  return new google.auth.OAuth2(
    process.env['GOOGLE_CLIENT_ID'],
    process.env['GOOGLE_CLIENT_SECRET'],
    process.env['REDIRECT_URI'],
  )
}

function parseDuration(iso: string): number {
  const m = iso.match(/PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?/)
  if (!m) return 0
  return (parseInt(m[1] ?? '0') * 3600) + (parseInt(m[2] ?? '0') * 60) + parseInt(m[3] ?? '0')
}

export class YouTubePlatform extends BasePlatform {
  private async getAuthClient(userId: string) {
    const tokens = getPlatformTokens(userId, 'youtube')
    if (!tokens) throw new Error(`No YouTube tokens for user ${userId}`)

    const auth = makeOAuth2Client()
    auth.setCredentials({
      access_token: tokens.accessToken,
      refresh_token: tokens.refreshToken,
      expiry_date: new Date(tokens.expiresAt).getTime(),
    })

    // Auto-refresh if needed
    auth.on('tokens', (newTokens) => {
      if (newTokens.access_token) {
        const expiresAt = newTokens.expiry_date
          ? new Date(newTokens.expiry_date).toISOString()
          : new Date(Date.now() + 3600_000).toISOString()
        updatePlatformTokens(userId, 'youtube', newTokens.access_token, expiresAt)
      }
    })

    return auth
  }

  async getRecentPosts(userId: string): Promise<NormalizedPost[]> {
    const auth = await this.getAuthClient(userId)
    const youtube = google.youtube({ version: 'v3', auth })

    // 1. Get uploads playlist ID
    const channelRes = await youtube.channels.list({
      part: ['contentDetails'],
      mine: true,
    })
    const uploadsPlaylistId =
      channelRes.data.items?.[0]?.contentDetails?.relatedPlaylists?.uploads
    if (!uploadsPlaylistId) return []

    // 2. Fetch playlist items (up to 50)
    const playlistRes = await youtube.playlistItems.list({
      part: ['contentDetails'],
      playlistId: uploadsPlaylistId,
      maxResults: 50,
    })
    const videoIds = (playlistRes.data.items ?? [])
      .map((i) => i.contentDetails?.videoId)
      .filter((id): id is string => Boolean(id))

    if (videoIds.length === 0) return []

    // 3. Batch-fetch video details (50 at a time)
    const posts: NormalizedPost[] = []
    for (let i = 0; i < videoIds.length; i += 50) {
      const batch = videoIds.slice(i, i + 50)
      const videosRes = await youtube.videos.list({
        part: ['snippet', 'statistics', 'contentDetails'],
        id: batch,
      })
      for (const v of videosRes.data.items ?? []) {
        const snippet = v.snippet ?? {}
        const stats = v.statistics ?? {}
        const content = v.contentDetails ?? {}

        const views = parseInt(stats.viewCount ?? '0')
        const likes = parseInt(stats.likeCount ?? '0')
        const comments = parseInt(stats.commentCount ?? '0')
        const durationSec = parseDuration(content.duration ?? '')
        const engagementRate = views > 0 ? ((likes + comments) / views) * 100 : 0

        posts.push({
          platform: 'youtube',
          postId: v.id ?? '',
          title: snippet.title ?? '',
          description: snippet.description ?? '',
          tags: snippet.tags ?? [],
          publishedAt: snippet.publishedAt ?? '',
          thumbnailUrl:
            snippet.thumbnails?.medium?.url ??
            snippet.thumbnails?.default?.url ??
            `https://img.youtube.com/vi/${v.id}/mqdefault.jpg`,
          metrics: {
            views,
            likes,
            comments,
            shares: 0,
            watchTimeSeconds: durationSec,
            impressions: 0,
            clickThroughRate: 0,
            engagementRate,
            averageViewPercentage: 0,
          },
        })
      }
    }

    return posts
  }

  async getPostAnalytics(userId: string, postId: string): Promise<PostMetrics> {
    const auth = await this.getAuthClient(userId)
    const youtubeAnalytics = google.youtubeAnalytics({ version: 'v2', auth })

    const endDate = new Date().toISOString().split('T')[0]!
    const startDate = new Date(Date.now() - 90 * 86400_000).toISOString().split('T')[0]!

    try {
      const res = await youtubeAnalytics.reports.query({
        ids: 'channel==MINE',
        startDate,
        endDate,
        metrics:
          'views,likes,comments,estimatedMinutesWatched,impressions,impressionClickThroughRate,averageViewPercentage',
        dimensions: 'video',
        filters: `video==${postId}`,
      })

      const row = res.data.rows?.[0]
      if (!row) {
        return {
          views: 0, likes: 0, comments: 0, shares: 0,
          watchTimeSeconds: 0, impressions: 0, clickThroughRate: 0,
          engagementRate: 0, averageViewPercentage: 0,
        }
      }

      const views = Number(row[1] ?? 0)
      const likes = Number(row[2] ?? 0)
      const comments = Number(row[3] ?? 0)
      return {
        views,
        likes,
        comments,
        shares: 0,
        watchTimeSeconds: Number(row[4] ?? 0) * 60,
        impressions: Number(row[5] ?? 0),
        clickThroughRate: Number(row[6] ?? 0),
        engagementRate: views > 0 ? ((likes + comments) / views) * 100 : 0,
        averageViewPercentage: Number(row[7] ?? 0),
      }
    } catch {
      return {
        views: 0, likes: 0, comments: 0, shares: 0,
        watchTimeSeconds: 0, impressions: 0, clickThroughRate: 0,
        engagementRate: 0, averageViewPercentage: 0,
      }
    }
  }

  async getChannelAnalytics(userId: string): Promise<ChannelAnalytics> {
    const defaultAnalytics: ChannelAnalytics = {
      totalViews: 0,
      totalWatchTime: 0,
      subscribersGained: 0,
      subscribersLost: 0,
      period: 'last_90_days',
    }

    try {
      const auth = await this.getAuthClient(userId)
      const youtubeAnalytics = google.youtubeAnalytics({ version: 'v2', auth })

      const endDate = new Date().toISOString().split('T')[0]!
      const startDate = new Date(Date.now() - 90 * 86400_000).toISOString().split('T')[0]!

      const res = await youtubeAnalytics.reports.query({
        ids: 'channel==MINE',
        startDate,
        endDate,
        metrics: 'views,estimatedMinutesWatched,subscribersGained,subscribersLost',
      })

      const row = res.data.rows?.[0] ?? [0, 0, 0, 0]
      return {
        totalViews: Number(row[0] ?? 0),
        totalWatchTime: Number(row[1] ?? 0),
        subscribersGained: Number(row[2] ?? 0),
        subscribersLost: Number(row[3] ?? 0),
        period: `${startDate} to ${endDate}`,
      }
    } catch (err) {
      console.warn('getChannelAnalytics failed, using defaults:', err instanceof Error ? err.message : err)
      return defaultAnalytics
    }
  }
}
